packages/nexus/src/channel.ts

```diff
@@ -0,0 +1,118 @@
+import { EventEmitter } from 'node:events'
+import TypedEmitter from 'typed-emitter'
+import { Data as WsData, WebSocket } from 'ws'
+import { NexusEvent, parseNexusEvent } from './events'
+
+export interface NexusChannelOptions {
+  reconnect?: boolean
+  reconnectDelay?: number
+}
+
+type NexusEvents = {
+  open: () => void
+  event: (evt: NexusEvent) => void
+  error: (err: Error) => void
+  reconnecting: (code: number, reason: string) => void
+  close: () => void
+}
+
+export type NexusEmitter = TypedEmitter<NexusEvents>
+
+export class NexusChannel extends (EventEmitter as new () => NexusEmitter) {
+  ws: WebSocket | null = null
+  url: string
+  reconnect: boolean
+  reconnectDelay: number
+  closed = false
+  reconnectTimer?: NodeJS.Timeout
+  bufferedAcks: number[] = []
+
+  constructor(url: string, options: NexusChannelOptions = {}) {
+    super()
+    this.url = url
+    this.reconnect = options.reconnect ?? true
+    this.reconnectDelay = options.reconnectDelay ?? 1000
+  }
+
+  ackEvent = (id: number): Promise<boolean> => {
+    return new Promise((resolve, reject) => {
+      if (this.ws && this.isConnected()) {
+        this.ws.send(JSON.stringify({ id }), (err) => {
+          if (err) {
+            reject(err)
+          } else {
+            resolve(true)
+          }
+        })
+      } else {
+        this.bufferedAcks.push(id)
+        resolve(true)
```

Comment:
I think there's a case to be made that this should resolve only once the buffered ack is processed.  I don't think it would be a very big change if we prefer that.